# Проект по SQL

Компания купила крупный сервис для чтения книг по подписке.

## Цель работы

Проанализировать базу данных.

## Задачи 

- Посчитаем, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определим автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками;
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


## Описание данных

Таблица books - содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица authors - содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица publishers - содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица ratings - содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица reviews - содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## Загрузка данных 

Импортируем библиотеки, загрузим данные и посмотрим на информацию о таблицах.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
books_query = '''
SELECT *
FROM books

'''
books = pd.io.sql.read_sql(books_query, con = engine)
display(books)
print(books.info())
print('Количество дубликатов:',books.duplicated().sum())

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Количество дубликатов: 0


In [3]:
authors_query = '''
SELECT *
FROM authors

'''
authors = pd.io.sql.read_sql(authors_query, con = engine)
display(authors)
print(authors.info())
print('Количество дубликатов:',authors.duplicated().sum())

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Количество дубликатов: 0


In [4]:
publishers_query = '''
SELECT *
FROM publishers

'''
publishers = pd.io.sql.read_sql(publishers_query, con = engine)
display(publishers)
print(publishers.info())
print('Количество дубликатов:', publishers.duplicated().sum())

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Количество дубликатов: 0


In [5]:
ratings_query = '''
SELECT *
FROM ratings

'''
ratings = pd.io.sql.read_sql(ratings_query, con = engine)
display(ratings)
print(ratings.info())
print('Количество дубликатов:', ratings.duplicated().sum())

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Количество дубликатов: 0


In [6]:
reviews_query = '''
SELECT *
FROM reviews 

'''
reviews = pd.io.sql.read_sql(reviews_query, con = engine)
display(reviews )
print(reviews .info())
print('Количество дубликатов:', reviews .duplicated().sum())

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Количество дубликатов: 0


### Вывод

Все библиотеки и таблицы загружены. Во всех таблицах нет пропусков и дубликатов.

## Задачи
### Задача 1. Посчитаем, сколько книг вышло после 1 января 2000 года.

In [7]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date >= '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine)

count
0    821

После 1 января 2000 года вышло 821 книга.

### Задача 2.  Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [8]:
query = '''
WITH
    count_review AS
    (SELECT book_id,
            COUNT(review_id) AS count_reviews
     FROM reviews
     GROUP BY book_id),
     
    avg_rating AS
    (SELECT book_id,
            ROUND(AVG(rating), 2) AS avg_ratings
     FROM ratings
     GROUP BY book_id)
      
    SELECT b.title,
           rc.count_reviews,
           ra.avg_ratings
    FROM books b
    LEFT JOIN count_review AS rc ON b.book_id = rc.book_id
    LEFT JOIN avg_rating AS ra ON b.book_id = ra.book_id
    ORDER BY ra.avg_ratings DESC;
    
'''
pd.io.sql.read_sql(query, con = engine)

title  count_reviews  \
0                        A Dirty Job (Grim Reaper  #1)            4.0   
1                             Tai-Pan (Asian Saga  #2)            2.0   
2              School's Out—Forever (Maximum Ride  #2)            3.0   
3                 Piercing the Darkness (Darkness  #2)            2.0   
4    The War of Art: Break Through the Blocks & Win...            2.0   
..                                                 ...            ...   
995  The World Is Flat: A Brief History of the Twen...            3.0   
996                  His Excellency: George Washington            2.0   
997                                              Junky            2.0   
998                                      Drowning Ruth            3.0   
999                               Harvesting the Heart            2.0   

     avg_ratings  
0           5.00  
1           5.00  
2           5.00  
3           5.00  
4           5.00  
..           ...  
995         2.25  
996         2.00  
997         2.00  
998         2.00  
999         1.50  

[1000 rows x 3 columns]

Для каждой книги посчитано количество обзоров и средняя оценка, по которой ранжировалась таблица.

### Задача 3.  Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [9]:
query = '''
SELECT p.publisher,
         COUNT(b.book_id)
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY  p.publisher
ORDER BY  COUNT(b.book_id) DESC
LIMIT 1
'''
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

Penguin Books выпустили 42 книги толще 50 страниц.

### Задача 4. Определим автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками.

In [10]:
query = '''
    SELECT a.author,
           ROUND(AVG(r.rating), 2) AS avg_rating 
    FROM books b
    LEFT JOIN authors a ON b.author_id = a.author_id
    LEFT JOIN ratings r ON b.book_id = r.book_id 
    WHERE b.book_id IN (SELECT book_id
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating) >= 50)
    GROUP BY a.author
    ORDER BY avg_rating DESC
    LIMIT(1);
             '''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой 4.29.

### Задача 5.  Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
query = '''
    WITH
    count_re AS
    (SELECT username,
            COUNT(review_id)
    FROM reviews
    GROUP BY username)
    
    SELECT ROUND(AVG(count), 2)
    FROM count_re
    WHERE username IN (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating) > 50);
             '''
pd.io.sql.read_sql(query, con = engine)

round
0  24.33

24,33 обзора в среднем от пользователей, которые поставили 50 оценок и более. 

## Вывод
Во всех таблицах нет пропусков и дубликатов.

После 1 января 2000 года вышло 821 книга. Для каждой книги посчитано количество обзоров и средняя оценка, по которой ранжировалась таблица. Penguin Books выпустили 42 книги толще 50 страниц. J.K. Rowling/Mary GrandPré - автор с самой высокой средней оценкой 4,29. 24,33 обзора в среднем от пользователей, которые поставили 50 оценок и более.